# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [35]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric
from sklearn.tree import DecisionTreeRegressor

In [36]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [37]:
x.shape[0]
array_sum = np.sum(x)
array_has_nan = np.isnan(array_sum)
print(array_has_nan)

False


In [38]:
type(x[0])

numpy.ndarray

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

<br><br><br>

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/BTVYXQ1.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [39]:
import random
from random import randrange
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 
    selecting_rows = np.random.choice(input_data.shape[0], int(np.floor(input_data.shape[0] * 0.6)), replace=False) 
    replacing_rows = np.random.choice(selecting_rows.shape[0], int(np.ceil(input_data.shape[0] * 0.4)), replace=False)
    selecting_columns = np.random.choice(input_data.shape[1], np.random.choice(np.arange(3, input_data.shape[1] + 1)), replace=False)
    sample_data = input_data[selecting_rows[:, None], selecting_columns]
    target_sample_data = target_data[selecting_rows]

    replicated_sample_data = sample_data[replacing_rows]
    target_replicated_sample_data = target_sample_data[replacing_rows]

    final_sample_data = np.vstack((sample_data, replicated_sample_data))
    final_target_data = np.vstack((target_sample_data.reshape(-1,1), target_replicated_sample_data.reshape(-1,1)))
    return final_sample_data,final_target_data,selecting_rows,selecting_columns
    # return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    #note please return as lists
a,b,c,d = generating_samples(x,y)    

In [40]:
len(b)

506

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [41]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)
print(a,b,c,d)

[[ 3.97    7.52    2.1398 ... 20.      0.647  89.4   ]
 [18.1     6.112   2.5091 ...  0.      0.77   81.3   ]
 [ 3.97    7.47    2.872  ... 20.      0.575  52.6   ]
 ...
 [18.1     7.061   3.4106 ...  0.      0.532  77.    ]
 [ 8.14    6.096   4.4619 ...  0.      0.538  84.5   ]
 [10.59    5.404   3.665  ...  0.      0.489  88.6   ]] [[43.1]
 [22.6]
 [43.5]
 [16.7]
 [23.1]
 [28. ]
 [19.1]
 [18.9]
 [22.5]
 [17.8]
 [ 8.3]
 [22. ]
 [33.1]
 [ 7.5]
 [23.4]
 [17.1]
 [14.9]
 [20.2]
 [19.5]
 [14.1]
 [19.2]
 [20.8]
 [23.1]
 [19.4]
 [20.4]
 [20.2]
 [20.6]
 [16. ]
 [ 7. ]
 [50. ]
 [33. ]
 [23.7]
 [10.9]
 [13.8]
 [24.4]
 [28.5]
 [18.9]
 [19.3]
 [23.3]
 [43.8]
 [21.2]
 [17.4]
 [28.2]
 [22.5]
 [22.2]
 [10.2]
 [12.7]
 [50. ]
 [20.9]
 [14.4]
 [20.3]
 [29.1]
 [21.4]
 [42.3]
 [16.2]
 [18.5]
 [20.6]
 [20.5]
 [22. ]
 [20. ]
 [ 6.3]
 [37. ]
 [10.5]
 [33.8]
 [ 8.5]
 [28.7]
 [19.1]
 [20.8]
 [18.6]
 [18.2]
 [32.9]
 [20.4]
 [20.6]
 [14.3]
 [18.8]
 [16.6]
 [ 8.4]
 [19.8]
 [27.1]
 [10.5]
 [ 7.2]
 [12.8]
 [13.3]


*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [44]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

for i in range(0,30):
    a,b,c,d = generating_samples(x,y)
    list_input_data.append(np.array(a))
    list_output_data.append(np.array(b))
    list_selected_row.append(c)
    list_selected_columns.append(d)

<font color='cyan'> <b>Grader function - 2 </b></font>

In [45]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [46]:
from sklearn.tree import DecisionTreeRegressor

list_of_all_model = []
for index ,value in enumerate(list_input_data):
    model = DecisionTreeRegressor(max_depth=None)
    model.fit(value,list_output_data[index])
    list_of_all_model.append(model)
    

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [47]:
Y_predicted_value = []
for value in x:
    #print(type(value))
    pred_y_list = []    
    for index, model in enumerate(list_of_all_model):
        pred_y = model.predict(value[list_selected_columns[index]].reshape(1, -1))
        pred_y_list.append(pred_y)
        #print(type(value[list_selected_columns[index]].reshape(1, -1)))
    median_pred_y = np.median(pred_y_list)
    Y_predicted_value.append(median_pred_y)
#print(predicted_y)
mse = mean_squared_error(y, Y_predicted_value)
print(mse)
        

0.2976136363636364


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [48]:
import warnings
warnings.filterwarnings("ignore")

sq_error=[]
for index, value in enumerate(x):
    all_pred_y = []
    for i, v in enumerate(list_output_data):
        if y[index] not in list_output_data[i]:
            pred_y = list_of_all_model[i].predict(value[list_selected_columns[i]].reshape(1, -1))
            all_pred_y.append(pred_y)               
    median_pred_y = np.median(np.asarray(all_pred_y))
    sq_error.append(np.square(y[index] - median_pred_y))

oob_val = np.nansum(sq_error) / len(x)
print(oob_val)

13.014100790513833


array([10,  3,  7,  1,  9,  5,  0, 12,  2])

# <font color='blue'><b>Task 2</b></font>

In [29]:
MSE_value = []
OOB_value = []
for index_value in range(0,35):
    list_input_data =[]
    list_output_data =[]
    list_selected_row= []
    list_selected_columns=[]

    for i in range(0,30):
        a,b,c,d = generating_samples(x,y)
        list_input_data.append(np.array(a))
        list_output_data.append(np.array(b))
        list_selected_row.append(c)
        list_selected_columns.append(d)
    list_of_all_model = []
    for index ,value in enumerate(list_input_data):
        model = DecisionTreeRegressor(max_depth=None)
        model.fit(value,list_output_data[index])
        list_of_all_model.append(model)
    Y_predicted_value = []
    for value in x:
        #print(type(value))
        pred_y_list = []    
        for index, model in enumerate(list_of_all_model):
            pred_y = model.predict(value[list_selected_columns[index]].reshape(1, -1))
            pred_y_list.append(pred_y)
            #print(type(value[list_selected_columns[index]].reshape(1, -1)))
        median_pred_y = np.median(pred_y_list)
        Y_predicted_value.append(median_pred_y)
    #print(predicted_y)
    mse = mean_squared_error(y, Y_predicted_value)
    MSE_value.append(mse)    
    sq_error=[]
    for index, value in enumerate(x):
        all_pred_y = []
        for i, v in enumerate(list_output_data):
            if y[index] not in list_output_data[i]:
                pred_y = list_of_all_model[i].predict(value[list_selected_columns[i]].reshape(1, -1))
                all_pred_y.append(pred_y)               
        median_pred_y = np.median(np.asarray(all_pred_y))
        sq_error.append(np.square(y[index] - median_pred_y))

    oob_val = np.nansum(sq_error) / len(x)
    OOB_value.append(oob_val)    

In [32]:
mean_MSE = np.asarray(MSE_value).mean()
std_MSE = np.asarray(MSE_value).std()

left_limit_MSE = np.round(mean_MSE - 2*(std_MSE/ np.sqrt(len(MSE_value))), 3)
right_limit_MSE = np.round(mean_MSE + 2*(std_MSE / np.sqrt(len(MSE_value))), 3)

print(mean_MSE,std_MSE,left_limit_MSE,right_limit_MSE)



0.0755311727252607 0.0542267855802621 0.057 0.094


In [33]:
mean_OOB = np.asarray(OOB_value).mean()
std_OOB = np.asarray(OOB_value).std()

left_limit_OOB = np.round(mean_OOB - 2*(std_OOB/ np.sqrt(len(OOB_value))), 3)
right_limit_OOB = np.round(mean_OOB + 2*(std_OOB / np.sqrt(len(OOB_value))), 3)

print(mean_OOB,std_OOB,left_limit_OOB,right_limit_OOB)


12.405236748883942 1.4784679761619712 11.905 12.905


# <font color='blue'><b>Task 3</b></font>

In [50]:
query_point =  [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60]
query_point=np.array(query_point)
pred_query_point = []
for index, model in enumerate(list_of_all_model):
        pred_query = model.predict(query_point[list_selected_columns[index]].reshape(1, -1))
        pred_query_point.append(pred_query)
median_query_point = np.median(pred_query_point) 
print(median_query_point)
    

18.5


<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

<br><br><br>

<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

# Task 1 Observation

# step 1



Used the boston house dataset to create 30 samples with row sampling and column sampling.

Row Sampling: As mentioned above selecting rows data is 60% of input data whereas remaining 40% data is replicated from selecting rows data.

Column Sampling: For column sampling, we have selected random columns which are greater than or equal to 3. 
Using vertical stack to concatenate the sample data and target sample data.

We have created 30 such samples which has different rows and different columns.

# step 2



Using the 30 samples we build a high variance DecisionTreeRegressor model with max_depth is None.

Built 30 such models and per model is per sample.

The MSE for every datapoint from input data is 0.2976136363636364

# step 3

Calculate the OOB score for every data point . The OOB score is 13.014100790513833

# Task 2 observations

Here, we have ran the 30 samples for 35 epochs and 30 models for each epoch. 

Calculated the mse and oob score for the input data and appended to a list after every epoch.

After getting the list of mse and oob score, I calculated the confidence interval for both mse and oob score. 

Confidence Interval for MSE is between 0.057 and  0.094

Confidence Interval for OOB score is between 11.905 and 12.905

# Task 3 Observation

The predicted value of the given query point is 18.5